clippy_lints/src/local_assigned_single_value.rs

```diff
@@ -0,0 +1,326 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::fn_has_unsatisfiable_preds;
+use clippy_utils::source::snippet_opt;
+use itertools::Itertools;
+use rustc_const_eval::interpret::Scalar;
+use rustc_data_structures::fx::FxHashSet;
+use rustc_hir::def_id::LocalDefId;
+use rustc_hir::{intravisit::FnKind, Body, FnDecl};
+use rustc_index::IndexVec;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::mir::{
+    self, interpret::ConstValue, visit::Visitor, Constant, Location, Mutability, Operand, Place, Rvalue,
+};
+use rustc_middle::mir::{AggregateKind, CastKind, PlaceElem, Terminator, TerminatorKind};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::source_map::Spanned;
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for locals that are always assigned the same value.
+    ///
+    /// ### Why is this bad?
+    /// It's almost always a typo. If not, it can be made immutable, or turned into a constant.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let mut x = 1;
+    /// x = 1;
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let x = 1;
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub LOCAL_ASSIGNED_SINGLE_VALUE,
+    correctness,
+    "disallows assigning locals many times with the same value"
+}
+declare_lint_pass!(LocalAssignedSingleValue => [LOCAL_ASSIGNED_SINGLE_VALUE]);
+
+impl LateLintPass<'_> for LocalAssignedSingleValue {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'_>,
+        _: FnKind<'_>,
+        _: &FnDecl<'_>,
+        _: &Body<'_>,
+        _: Span,
+        def_id: LocalDefId,
+    ) {
+        // Building MIR for `fn`s with unsatisfiable preds results in ICE.
+        if fn_has_unsatisfiable_preds(cx, def_id.to_def_id()) {
+            return;
+        }
+
+        let mir = cx.tcx.optimized_mir(def_id.to_def_id());
+        let mut v = V {
+            body: mir,
+            cx,
+            map: mir.local_decls.iter().map(|_| LocalUsageValues::default()).collect(),
```

Comment:
The map should be initialized to a state such that only user bound mutable variables check the assigned value. i.e. `matches!(l.local_info, LocalInfo::User(_)) && l.mutability == Mutability::Mut`)

Reply:
Well, `local_info` is actually `ClearCrossCrate` now so we can't. It's not available past borrowck and friends and results in an ICE if used in clippy.

We could perhaps use `var_debug_info` or something.